src/backend/distributed/planner/shard_pruning.c

```diff
@@ -825,6 +852,45 @@ AddPartitionKeyRestrictionToInstance(ClauseWalkerContext *context, OpExpr *opCla
 }
 
 
+/*
+ * Sometimes PostgreSQL chooses to try to wrap our Var in a coercion rather
+ * than the Const; to deal with this, we strip the coercions from both and
+ * manually coerce the Const into the type of our partition column. It is
+ * conceivable that in some instances, this may not be possible; in those cases
+ * we will simply fail to prune partitions based on this clause.
+ */
+static Const *
+TransformPartitionRestrictionValue(Var *partitionColumn, Node *restrictionValue)
+{
+	Node *transformedValue = coerce_to_target_type(NULL, restrictionValue,
+												   exprType(restrictionValue),
```

Comment:
I'm wondering whether should we have a a short circuit for `if exprType(restrictionValue) == partitionColumn->vartype`? That's going to be the vast majority of the time and I'd like to avoid any extra overhead.

I can see that `coerce_to_target_type ` does some very little things if that's already the case. But, given that this function will be called too many times on busy servers, I'd have a slight preference to have the short-circuit (or maybe run a quick select-only benchmark to see if it has any noticeable overhead or not.).

(And, that'd probably easier to follow: `If the types are the same return else transform`)

Reply:
Sure, this seems reasonable. Also, I think I probably am pushed toward blindly calling a function and assigning it during declaration because of the weird style requirement around that.

Reply:
OK, I actually put the check _before_ the call to `TransformPartitionRestrictionValue`, as that way we can actually save the function call overhead, too. I think that this will be the _overwhelmingly_ common case, and as such branch prediction should _probably_ make all of this moot, but just in case.